File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/digitaltwin/element/ElementListType.java
package org.opcfoundation.webserver.digitaltwin.element;

import org.eclipse.milo.opcua.sdk.core.AccessLevel;
import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.opcfoundation.webserver.addressspace.nodes.*;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaObjectTypes;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.callback.ElementListCallback;
import org.opcfoundation.webserver.types.ObjectElementDescriptor;
import org.opcfoundation.webserver.types.PropertyElementDescriptor;
import org.opcfoundation.webserver.types.ServiceContext;
import org.opcfoundation.webserver.types.UaBrowseAdditionalInfo;
import org.opcfoundation.webserver.types.UaChildId;
import org.opcfoundation.webserver.types.UaReferenceDescriptor;
import org.opcfoundation.webserver.types.message.*;
import org.opcfoundation.webserver.types.message.digitaltwin.*;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public abstract class ElementListType extends ElementType implements ElementListCallback {
    public ElementListType(String typeId,
                           LocalizedText displayName,
                           DigitalTwinSpace namespace)
    {
        super(
                typeId,
                displayName,
                UaObjectTypes.ElementType,
                namespace);
    }

    @Override
    public final CompletableFuture<ReadObjectAttributeResponse> onReadObjectAttributes(ReadObjectAttributeRequest request)
    {
        UaObject instanceDeclaration = request.getObjectId().getInstance();

        if (null == instanceDeclaration)
        {
            ServiceContext context = new ServiceContext(request.getObjectId());
            GetDescriptorRequest getDescriptorRequest = new GetDescriptorRequest(context);

            return onGetDescriptor(getDescriptorRequest).thenApply(response -> {
                return new ReadObjectAttributeResponse(response.getDisplayName(), response.getDescription());
            });
        } else {
            ReadObjectAttributeResponse response = new ReadObjectAttributeResponse(
                    instanceDeclaration.displayName(),
                    instanceDeclaration.description());

            return CompletableFuture.completedFuture(new ReadObjectAttributeResponse(response.getDisplayName(), response.getDescription()));
        }
    }

    @Override
    public final CompletableFuture<BrowseObjectResponse> onBrowseObjectChildren(BrowseObjectRequest request)
    {
        ServiceContext context = new ServiceContext(request.getObjectId());

        if (request.getAdditionalInfo().isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK) &&
                supportObjectElementList())
        {
            GetObjectElementListRequest getObjectElementListRequest = new GetObjectElementListRequest(
                    context,
                    request.getAdditionalInfo().getMaxReferencesPerNode(),
                    request.getAdditionalInfo().getReferenceOffset());

            return onGetObjectElementList(getObjectElementListRequest).
                    thenApply(this::processBrowseObjectChildrenResponse);
        } else if (request.getAdditionalInfo().isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK) &&
                supportPropertyElementList() && !request.getBrowseDescription().getReferenceTypeId().equals(NodeIds.HasProperty))
        {
            GetPropertyElementListRequest getPropertyElementListRequest = new GetPropertyElementListRequest(
                    context,
                    request.getAdditionalInfo().getMaxReferencesPerNode(),
                    request.getAdditionalInfo().getReferenceOffset());

            return onGetPropertyElementList(getPropertyElementListRequest).
                    thenApply(this::processBrowseObjectChildrenResponse);
        } else {
            return CompletableFuture.completedFuture(new BrowseObjectResponse(new ArrayList<>(), false));
        }
    }

    @Override
    public final CompletableFuture<BrowseMemberResponse> onBrowseMemberChildren(BrowseMemberRequest request)
    {
        UaNode hasComponentType = nodeManager.getNode(NodeIds.HasComponent);

        if (null == hasComponentType || hasComponentType.nodeClass() != NodeClass.ReferenceType)
        {
            throw new UaRuntimeException(StatusCodes.Bad_InternalError);
        }

        ServiceContext context = new ServiceContext(request.getObjectId());
        GetPropertySubElementsRequest getPropertySubElementsRequest= new GetPropertySubElementsRequest(
                context,
                request.getChildId());

        return onGetPropertySubElements(getPropertySubElementsRequest).
                thenApply(this::processBrowseMemberChildren);
    }

    @Override
    public final CompletableFuture<ReadMemberAttributeResponse> onReadMemberAttributes(ReadMemberAttributeRequest request)
    {
        ServiceContext context = new ServiceContext(request.getObjectId());
        GetPropertyDescriptorRequest getPropertyDescriptorRequest = new GetPropertyDescriptorRequest(
                context,
                request.getChildId().getId(),
                request.getChildId().getSubElementName());

        return onGetPropertyDescriptor(getPropertyDescriptorRequest).thenApply(this::processReadMemberAttributeResponse);
    }

    @Override
    public final CompletableFuture<ReadVariableValueResponse> onReadVariablesValue(ReadVariableValueRequest request)
    {
        final Set<String> propertyIds = new HashSet<>();
        final Set<UaChildId> subPropertyIds = new HashSet<>();

        for (UaChildId item: request.getVariableIds())
        {
            if (null == item.getSubElementName())
            {
                propertyIds.add(item.getId());
            } else {
                subPropertyIds.add(item);
            }
        }

        ServiceContext context = new ServiceContext(request.getObjectId());
        ReadPropertyListValueRequest readPropertyValuesRequest = new ReadPropertyListValueRequest(
                context,
                propertyIds,
                subPropertyIds);

        return onReadPropertyValues(readPropertyValuesRequest).
                thenApply(readPropertyValuesResponse -> {
                    return new ReadVariableValueResponse(readPropertyValuesResponse.getResults()); });
    }

    @Override
    public final CompletableFuture<WriteVariableValueResponse> onWriteVariablesValue(WriteVariableValueRequest request)
    {
        Map<String, Variant> propertyIdAndValues = new HashMap<>();
        Map<UaChildId, Variant> subPropertyIdsAndValues = new HashMap<>();

        for (Map.Entry<UaChildId, Variant> item : request.getVariableValues().entrySet())
        {
            UaChildId childId = item.getKey();

            if (null == childId.getSubElementName())
            {
                UaNode memberNode = getMember(childId.getId());
                if (null == memberNode || NodeClass.Variable != memberNode.nodeClass()) continue;

                UaVariable variableToWrite = (UaVariable) memberNode;
                if ((AccessLevel.CurrentWrite.getValue() & variableToWrite.accessLevel()) == 0) continue;

                propertyIdAndValues.put(childId.getId(), item.getValue());
            } else {
                subPropertyIdsAndValues.put(childId, item.getValue());
            }
        }

        if (propertyIdAndValues.isEmpty() && subPropertyIdsAndValues.isEmpty())
        {
            return CompletableFuture.completedFuture(new WriteVariableValueResponse());
        }

        ServiceContext context = new ServiceContext(request.getObjectId());
        WritePropertyListValuesRequest writePropertyValuesRequest = new WritePropertyListValuesRequest(
                context,
                propertyIdAndValues,
                subPropertyIdsAndValues);

        return onWritePropertyValues(writePropertyValuesRequest)
                .thenApply(response -> {
                    return new WriteVariableValueResponse(response.getResults());
                });
    }

    private BrowseObjectResponse processBrowseObjectChildrenResponse(GetObjectElementListResponse response)
    {
        List<UaReferenceDescriptor> childDescriptors = new ArrayList<>();

        for (ObjectElementDescriptor item: response.getElements())
        {
            UaReferenceDescriptor descriptor = new UaReferenceDescriptor(
                    item.getId(),
                    NodeClass.Object,
                    item.getId(),
                    item.getDisplayName(),
                    item.getTypeId(),
                    NodeIds.HasComponent,
                    true);

            childDescriptors.add(descriptor);
        }

        return new BrowseObjectResponse(
                childDescriptors,
                response.containsMoreData(),
                UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK);
    }

    private BrowseObjectResponse processBrowseObjectChildrenResponse(GetPropertyElementListResponse response)
    {
        List<UaReferenceDescriptor> childDescriptors = new ArrayList<>();

        for (PropertyElementDescriptor item: response.getElements())
        {
            UaReferenceDescriptor descriptor = new UaReferenceDescriptor(
                    item.getId(),
                    NodeClass.Variable,
                    item.getId(),
                    item.getDisplayName(),
                    item.getTypeId(),
                    NodeIds.HasComponent,
                    true);

            childDescriptors.add(descriptor);
        }

        return new BrowseObjectResponse(
                childDescriptors,
                response.containsMoreData(),
                UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK);
    }

    private BrowseMemberResponse processBrowseMemberChildren(GetPropertySubElementsResponse response)
    {
        List<UaReferenceDescriptor> childDescriptors = new ArrayList<>();

        for (String item: response.getSubElementNames())
        {
            UaReferenceDescriptor descriptor = new UaReferenceDescriptor(
                    item,
                    NodeClass.Variable,
                    item,
                    new LocalizedText(item),
                    NodeIds.PropertyType,
                    NodeIds.HasProperty,
                    true);

            childDescriptors.add(descriptor);
        }

        return new BrowseMemberResponse(childDescriptors);
    }

    private ReadMemberAttributeResponse processReadMemberAttributeResponse(GetPropertyDescriptorResponse response)
    {
        return new ReadMemberAttributeResponse(
                NodeClass.Variable,
                response.getDisplayName(),
                response.getDescription(),
                response.getDataTypeId(),
                response.getValueRank(),
                response.getAccessLevel(),
                response.getHistorizing());
    }
}

File: opcua-webservice-node/source/src/server/digitaltwin/element/ElementListType.ts
import { NodeClass, StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaModellingRule, ReferenceTypeIds, UaError, UaLocalizedText, UaNodeId, UaVariant, VariableTypeIds } from "opcua-webapi-ts";
import { UaObjectTypes } from "../../addressspace/nodes/builtin";
import {
    BrowseMemberRequest,
    BrowseMemberResponse,
    BrowseObjectRequest,
    BrowseObjectResponse,
    GetDescriptorRequest,
    GetObjectElementListRequest,
    GetObjectElementListResponse,
    GetPropertyDescriptorRequest,
    GetPropertyDescriptorResponse,
    GetPropertyElementListRequest,
    GetPropertyElementListResponse,
    GetPropertySubElementsRequest,
    GetPropertySubElementsResponse,
    ReadHistoryDataRequest,
    ReadHistoryDataResponse,
    ReadMemberAttributeRequest,
    ReadMemberAttributeResponse,
    ReadPropertyHistoryValuesRequest,
    ReadPropertyHistoryValuesResponse,
    ReadPropertyListValueRequest,
    ReadPropertyListValueResponse,
    ReadVariableValueRequest,
    ReadVariableValueResponse,
    WritePropertyListValuesRequest,
    WritePropertyListValuesResponse,
    WriteVariableValueRequest,
    WriteVariableValueResponse,
} from "../../service/message";
import { UaBrowseAdditionalInfo, UaChildId, UaReferenceDescriptor } from "../../types";
import { ObjectServiceContext } from "../../types/digitaltwin";
import { DigitalTwinSpace } from "../DigitalTwinSpace";
import { ElementType } from "./ElementType";

export abstract class ElementListType extends ElementType {
    constructor(
        typeId: string,
        displayName: UaLocalizedText,
        twinSpace: DigitalTwinSpace) {
        super(typeId, displayName, UaObjectTypes.ElementListType, twinSpace);
    }

    /**
     * Set the element type can be added to this element list.
     */
    mayAdd(type: ElementType): void {
        const newObject = this.addObjectNode(type.name, type.displayName, type);       
        newObject.setModellingRule(UaModellingRule.PlaceHolder);
    }

    /**
     * Optional override point to enable dynamic object-element listing.
     */
    supportObjectElementList(): boolean { return false; }

    /**
     * Override in subclasses to return dynamic object elements.
     */
    async onGetObjectElementList(request: GetObjectElementListRequest): Promise<GetObjectElementListResponse>
    {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }

    /**
     * Optional override point to enable dynamic property-element listing.
     */
    supportPropertyElementList(): boolean { return false; }

    /**
     * Override in subclasses to return the property element list.
     */
    async onGetPropertyElementList(request: GetPropertyElementListRequest): Promise<GetPropertyElementListResponse>
    {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }

    /**
     * Override in subclasses to return a property descriptor.
     */
    async onGetPropertyDescriptor(request: GetPropertyDescriptorRequest): Promise<GetPropertyDescriptorResponse>
    {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }

    /**
     * Override in subclasses to read property values.
     */
    async onReadPropertyValues(request: ReadPropertyListValueRequest): Promise<ReadPropertyListValueResponse>
    {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }
    
    /**
     * Override in subclasses to write property values.
     */
    async onWritePropertyValues(request: WritePropertyListValuesRequest): Promise<WritePropertyListValuesResponse>
    {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }

    /**
     * Override in subclasses to read historical property values.
     */
    async onReadPropertyHistoryValues(request: ReadPropertyHistoryValuesRequest): Promise<ReadPropertyHistoryValuesResponse>
    {
        throw new UaError(makeUaStatusCode(StatusCodes.BadNotImplemented));
    }

    /**
     * Optional override point to return sub-elements of a property.
     */
    async onGetPropertySubElements(request: GetPropertySubElementsRequest): Promise<GetPropertySubElementsResponse>
    {
        return new GetPropertySubElementsResponse();
    }     

    /**
     * Internal framework callback used by the base type to get the reference type id for this repository.
     * Do not call or override this method directly.
     */
    supportedReferenceType(): UaNodeId {
        return UaNodeId.from(ReferenceTypeIds.HasComponent);
    }

    /**
     * Internal framework callback used by the base type to browse child nodes.
     * Do not call or override this method directly.
     */
    override async onBrowseObject(request: BrowseObjectRequest): Promise<BrowseObjectResponse> {
        const context = new ObjectServiceContext(request.objectId);

        request.additionalInfo.taskComplete(UaBrowseAdditionalInfo.GET_CHILD_METHOD_TASK);

        if (request.additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK) &&
            this.supportObjectElementList()) {
            const response = await this.onGetObjectElementList(
                new GetObjectElementListRequest(
                    context,
                    request.additionalInfo.maxReferencesPerNode,
                    request.additionalInfo.referenceOffset,
                ),
            );

            return this.processBrowseObjectResponse(response,request.additionalInfo);
        }

        if (request.additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK) &&
            this.supportPropertyElementList()) {
            const response = await this.onGetPropertyElementList(
                new GetPropertyElementListRequest(
                    context,
                    request.additionalInfo.maxReferencesPerNode,
                    request.additionalInfo.referenceOffset,
                ),
            );

            return this.processBrowsePropertyResponse(response,request.additionalInfo);
        }

        return new BrowseObjectResponse([]);
    }

    /**
     * Internal framework callback used by the base type to browse member children.
     * Do not call or override this method directly.
     */
    override async onBrowseMember(request: BrowseMemberRequest): Promise<BrowseMemberResponse> {
        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onGetPropertySubElements(
            new GetPropertySubElementsRequest(context, request.childId),
        );

        return this.processBrowseMember(response);
    }

    /**
     * Internal framework callback used by the base type to read member attributes.
     * Do not call or override this method directly.
     */
    override async onReadMemberAttributes(request: ReadMemberAttributeRequest): Promise<ReadMemberAttributeResponse> {
        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onGetPropertyDescriptor(
            new GetPropertyDescriptorRequest(
                context,
                request.childId.id,
                request.childId.subElementName,
            ),
        );

        return this.processReadMemberAttributeResponse(request.childId, response);
    }

    /**
     * Internal framework callback used by the base type to read variable values.
     * Do not call or override this method directly.
     */
    override async onReadVariablesValue(request: ReadVariableValueRequest): Promise<ReadVariableValueResponse> {
        const propertyIds = new Set<string>();
        const subPropertyIds = new Set<UaChildId>();

        for (const item of request.variableIds) {
            const childId = UaChildId.fromString(item);
            if (childId.subElementName === null) {
                propertyIds.add(childId.id);
            } else {
                subPropertyIds.add(childId);
            }
        }

        const context = new ObjectServiceContext(request.objectId);
        const readPropertyValuesResponse = await this.onReadPropertyValues(
            new ReadPropertyListValueRequest(context, propertyIds, subPropertyIds),
        );

        const response = new ReadVariableValueResponse();
        for (const [key, value] of readPropertyValuesResponse.results) {
            response.results.set(key, value);
        }
        return response;
    }

    /**
     * Internal framework callback used by the base type to write variable values.
     * Do not call or override this method directly.
     */
    override async onWriteVariablesValue(request: WriteVariableValueRequest): Promise<WriteVariableValueResponse> {
        const propertyIdAndValues = new Map<string, UaVariant>();
        const subPropertyIdsAndValues = new Map<UaChildId, UaVariant>();

        for (const [childId, value] of request.variableValues) {
            const variableId = UaChildId.fromString(childId);
            if (variableId.subElementName === null) {
                propertyIdAndValues.set(variableId.id, value);
            } else {
                subPropertyIdsAndValues.set(variableId, value);
            }
        }

        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onWritePropertyValues(
            new WritePropertyListValuesRequest(context, propertyIdAndValues, subPropertyIdsAndValues),
        );

        return new WriteVariableValueResponse(response.results);
    }

    /**
     * Internal framework callback used by the base type to read history data.
     * Do not call or override this method directly.
     */
    override async onReadHistoryData(request: ReadHistoryDataRequest): Promise<ReadHistoryDataResponse> {
        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onReadPropertyHistoryValues(
            new ReadPropertyHistoryValuesRequest(
                context,
                request.childId.id,
                request.details,
            ),
        );

        return this.processReadHistoryValueResponse(response);
    }

    private processBrowseObjectResponse(
        response: GetObjectElementListResponse,
        additionalInfo: UaBrowseAdditionalInfo): BrowseObjectResponse {
        const childDescriptors: Array<UaReferenceDescriptor> = [];
        const referenceType = this.supportedReferenceType();

        for (const item of response.elements) {
            childDescriptors.push(
                new UaReferenceDescriptor(
                    item.id,
                    NodeClass.Object,
                    item.id,
                    item.displayName,
                    item.typeId,
                    referenceType
                ),
            );
        }

        let taskMask = (response.containsMoreData) ? additionalInfo.taskCheckListMasks : additionalInfo.taskCheckListMasks & ~UaBrowseAdditionalInfo.GET_RELATED_OBJECT_TASK;
        let offset = (response.containsMoreData) ? additionalInfo.referenceOffset + response.elements.length : 0;
        return new BrowseObjectResponse(childDescriptors, taskMask, offset);
    }

    private processBrowsePropertyResponse(
        response: GetPropertyElementListResponse,
        additionalInfo: UaBrowseAdditionalInfo): BrowseObjectResponse {
        const childDescriptors: Array<UaReferenceDescriptor> = [];
        const referenceType = (response.isProperty) ? UaNodeId.from(ReferenceTypeIds.HasProperty) : UaNodeId.from(ReferenceTypeIds.HasComponent);

        for (const item of response.elements) {
            childDescriptors.push(
                new UaReferenceDescriptor(
                    item.id,
                    NodeClass.Variable,
                    item.id,
                    item.displayName,
                    item.typeId,
                    referenceType
                ),
            );
        }

        let taskMask = (response.containsMoreData) ? additionalInfo.taskCheckListMasks : additionalInfo.taskCheckListMasks & ~UaBrowseAdditionalInfo.GET_CHILD_VARIABLE_TASK;
        let offset = (response.containsMoreData) ? additionalInfo.referenceOffset + response.elements.length : 0;
        return new BrowseObjectResponse(childDescriptors,taskMask,offset);
    }

    private processBrowseMember(response: GetPropertySubElementsResponse): BrowseMemberResponse {
        const childDescriptors: Array<UaReferenceDescriptor> = [];
        const variableType = UaNodeId.from(VariableTypeIds.PropertyType);
        const referenceType = UaNodeId.from(ReferenceTypeIds.HasProperty);

        for (const item of response.subElementNames) {
            childDescriptors.push(
                new UaReferenceDescriptor(
                    item,
                    NodeClass.Variable,
                    item,
                    new UaLocalizedText(item),
                    variableType,
                    referenceType
                ),
            );
        }

        return new BrowseMemberResponse(childDescriptors);
    }

    private processReadMemberAttributeResponse(childId: UaChildId, response: GetPropertyDescriptorResponse): ReadMemberAttributeResponse {
        return new ReadMemberAttributeResponse(
            NodeClass.Variable,
            childId.subElementName === null ? childId.id : childId.subElementName,
            response.displayName,
            response.description,
            response.dataTypeId,
            response.valueRank,
            response.accessLevel,
            response.historizing,
            null,
        );
    }

    private processReadHistoryValueResponse(response: ReadPropertyHistoryValuesResponse): ReadHistoryDataResponse {
        return new ReadHistoryDataResponse(response.dataValues, response.containsMoreData);
    }
}